app/src/main/kotlin/io/homeassistant/companion/android/widgets/BaseGlanceEntityWidgetReceiver.kt

```diff
@@ -0,0 +1,284 @@
+ * - Make sur to invoke [registerReceiver] to register actions to manage widget updates effectively.
+ * - Even if we are watching for entity changes in the receiver the data received cannot be sent to the widget directly,
+ *   if you need to send data to the widget you need to update the [DAO] and call [update].
+ * - The changes are watch only after [Intent.ACTION_SCREEN_ON]. It means that on startup the widget won't watch for changes
+ *   until the screen is turn on and off.
+ *
+ * ### Usage:
+ * Extend this class and implement the abstract method `getWidgetEntitiesByServer` to provide the mapping of widget IDs
+ * to their associated entities.
+ *
+ * ### Example:
+ * ```kotlin
+ * class MyWidgetReceiver : BaseGlanceWidgetReceiver<MyWidgetDao>() {
+ *     override suspend fun getWidgetEntitiesByServer(context: Context): Map<Int, EntitiesPerServer> {
+ *         // Return a map of widget IDs to their associated entities
+ *     }
+ * }
+ * ```
+ *
+ * Register the widget in the [io.homeassistant.companion.android.HomeAssistantApplication]:
+ * ```kotlin
+ * // In io.homeassistant.companion.android.HomeAssistantApplication
+ * override fun onCreate() {
+ *     super.onCreate()
+ *     ...
+ *     MyWidgetReceiver().registerReceiver(this)
+ * }
+ * ```
+ *
+ * ### Implementation Details:
+ * - **Coroutine Scope**: A shared [CoroutineScope] is used to manage asynchronous tasks for all widgets of a given class. It uses
+ *   [SupervisorJob] to ensure that one failing job does not cancel others.
+ * - **Entity updates**: Subscriptions to entity updates are managed per widget. When a widget is deleted, its
+ *   subscriptions are canceled, and its data is removed from the database.
+ *
+ * @param DAO The type of the DAO used for managing widget data in the database. It must implement [WidgetDao] and
+ *            be injectable with Hilt.
+ */
+abstract class BaseGlanceEntityWidgetReceiver<DAO : WidgetDao> @VisibleForTesting constructor(
+    private val widgetScopeProvider: () -> CoroutineScope,
+    private val glanceManagerProvider: (Context) -> GlanceAppWidgetManager,
+) : GlanceAppWidgetReceiver() {
+
+    constructor() : this(newCoroutineScopeProvider(), { GlanceAppWidgetManager(it) })
+
+    @Inject
+    lateinit var dao: DAO
+
+    @Inject
+    lateinit var serverManager: ServerManager
+
+    private var widgetScope: CoroutineScope = widgetScopeProvider()
+
+    private val widgetJobs = mutableMapOf<Int, Job>()
+
+    // Helper to identify the widget in the logs
+    private val widgetClassName by lazy { glanceAppWidget.javaClass.name }
+
+    override fun onReceive(context: Context, intent: Intent) {
+        val appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
+        Timber.v("Received intent action = ${intent.action} for widget = $appWidgetId ($widgetClassName) $this")
```

Comment:
Do we need debug log statements at the start of most functions?

The documentation you wrote suggests not using excessive logging and it should add value. While it may be true for development once accepted most of these feel less relevant to me.

Reply:
I kept this under verbose to see if we properly receive events for a given Intent. I think this one should be kept. I will cleanup the rest, but here being on a baseClass I think it is nice to be a bit verbose for people using the class in future development.